src/main/java/com/spotify/sparkey/extra/ThreadLocalWithFallback.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2022 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.spotify.sparkey.extra;
+
+import java.lang.ref.PhantomReference;
+import java.lang.ref.ReferenceQueue;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+/**
+ * Since Java 17, using ThreadLocal may not behave as expected
+ * when being called from ForkJoinPool.commonPool().
+ * The common pool will clear all values for ThreadLocal after finishing a task,
+ * so it can not be used for retaining expensive objects on the same thread across
+ * multiple tasks.
+ *
+ * This subclass will instead let values survive by using a ConcurrentHashMap
+ * based on Thread Id which is stable as long as the thread lives. This map is also cleaned up by
+ * detecting thread death using a WeakReference.
+ */
+class ThreadLocalWithFallback<T> extends ThreadLocal<T> {
+  private final Map<Long, T> fallback = new ConcurrentHashMap<>();
+  private final Supplier<? extends T> supplier;
+  private final Consumer<? super T> closer;
+
+  private final ReferenceQueue<Thread> deadThreads = new ReferenceQueue<>();
+  // Used to avoid creating more weak references than necessary
+  private final Map<Long, RefWithThreadId> activeThreads = new ConcurrentHashMap<>();
+
+  private ThreadLocalWithFallback(Supplier<? extends T> supplier, Consumer<? super T> closer) {
+    this.supplier = supplier;
+    this.closer = closer;
+  }
+
+  @Override
+  protected T initialValue() {
+    final Thread thread = Thread.currentThread();
+    final long threadId = thread.getId();
+    final T fallbackValue = fallback.get(threadId);
+    if (fallbackValue != null) {
+      // This case is typically only reachable on Java 17+ using ForkJoinPool.commonPool()
+      return fallbackValue;
+    }
+
+    final T value = Objects.requireNonNull(supplier.get());
+    setFallback(value, thread, threadId);
+    return value;
```

Comment:
nitty-pick: Can be simplified with `computeIfAbsent`

```suggestion
    return fallback.computeIfAbsent(tid -> {
      final T value = Objects.requireNonNull(supplier.get());
      setFallback(value, thread, threadId);
      return value;
    });
```

Reply:
Might be nice to avoid the extra object churn since the lambda references values on the outside (supplier, thread, threadId)